rpm/bundle-rust-downstream

```diff
@@ -0,0 +1,142 @@
+#!/usr/bin/python3
+
+# --- BEGIN COPYRIGHT BLOCK ---
+# Copyright (C) 2021 Red Hat, Inc.
+# All rights reserved.
+#
+# License: GPL (version 3 or any later version).
+# See LICENSE for details.
+# --- END COPYRIGHT BLOCK ---
+#
+# PYTHON_ARGCOMPLETE_OK
+
+import os
+import sys
+import signal
+import argparse
+import argcomplete
+import shutil
+import toml
+from lib389.cli_base import setup_script_logger
+
+SPECFILE_COMMENT_LINE = 'Bundled cargo crates list'
+START_LINE = f"# {SPECFILE_COMMENT_LINE} - START\n"
+END_LINE = f"# {SPECFILE_COMMENT_LINE} - END\n"
+
+
+parser = argparse.ArgumentParser(
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    description="""Add 'Provides: bundled(crate(foo)) = version' to a Fedora based specfile. 
+Additionally, add a helper comment with a comulated License metainfo which is based on Cargo.lock file content.""")
+
+parser.add_argument('-v', '--verbose',
+                    help="Display verbose operation tracing during command execution",
+                    action='store_true', default=False)
+
+parser.add_argument('cargo_lock_file',
+                    help="The path to Cargo.lock file.")
+parser.add_argument('spec_file',
+                    help="The path to spec file that will be modified.")
+parser.add_argument('--backup-specfile',
+                    help="Make a backup of the downstream specfile.",
+                    action='store_true', default=False)
+
+
+# handle a control-c gracefully
+def signal_handler(signal, frame):
+    print('\n\nExiting...')
+    sys.exit(0)
+
+
+def backup_specfile(spec_file):
+    log.info(f"Backing up file {spec_file} to {spec_file}.backup")
+    shutil.copy2(spec_file, f"{spec_file}.backup")
+
+
+def clean_specfile(spec_file):
+    result = []
+    remove_lines = False
+    cleaned = False
+    with open(spec_file, "r") as file:
+        contents = file.readlines()
+
+    log.info(f"Remove '{SPECFILE_COMMENT_LINE}' content from {spec_file}")
+    for line in contents:
+        if line == START_LINE:
+            remove_lines = True
+            log.debug(f"Remove '{START_LINE}' from {spec_file}")
+        elif line == END_LINE:
+            remove_lines = False
+            cleaned = True
+            log.debug(f"Remove '{END_LINE}' from {spec_file}")
+        elif not remove_lines:
+            result.append(line)
+        else:
+            log.debug(f"Remove '{line}' from {spec_file}")
+
+    with open(spec_file, "w") as file:
+        file.writelines(result)
+    return cleaned
+
+
+def write_provides_bundled_crate(cargo_lock_file, spec_file, cleaned):
+    # Generate 'Provides' out of cargo_lock_file
+    with open(cargo_lock_file, "r") as file:
+        contents = file.read()
+    data = toml.loads(contents)
+    provides_lines = []
+    for package in data["package"]:
+        provides_lines.append(f"Provides:  bundled(crate({package['name']})) = {package['version']}\n")
```

Comment:
subtle issue, but some crate dependencies have multiple hyphens in the name which can cause spec file errors as the parser thinks the version is incorrect and has multiple hyphens. My tool has a work around here:

https://github.com/Firstyear/cargo-lock2rpmprovides/blob/main/src/main.rs#L176

Reply:
rust2rpm has normalization / mapping functions from cargo/semver into RPM-compatible version strings:
https://pagure.io/fedora-rust/rust2rpm/blob/master/f/rust2rpm/metadata.py

Using those would make the bundled Provides use the same version strings as the crates packaged as "native" RPM pacakges.

Reply:
Hmm I'm not clear on how this code would handle the version string for 'wasi'. It has a version string of '0.9.0+wasi-snapshot-preview1' which is a dependency of getrandom -> ahash -> concread. The version string here is valid for rust, but rpm was having issues at it contains multiple hypens in it's current form. How would you normalise or parse this?

Reply:
You're right, it appears that it doesn't handle this case. We don't have wasi packaged yet, which is probably why this bug went unnoticed. However, for other packages with "+" in the version string (like cargo-c, zstd-sys, libgit2-sys, etc) we normalize these downstream with patches (since the +suffix is not strictly version information, but "context" information).

Anyway, reported for rust2rpm: https://pagure.io/fedora-rust/rust2rpm/issue/124

Reply:
Commented on the issue/subscribed. It would be good to know what you plan to do to normalise here so I can keep cargo-lock2rpmprovides consistent.

At the very least, for now I think we can just normalise it by hand as we generate the provides line until we have this sorted/fixed.